src/libraries/sortedLinkedList.se

```diff
@@ -0,0 +1,156 @@
+data comparor
+data collection[](exists, prev, next)
+data count
+data head
+data tail
+data owner
+data initialized
+
+inset('../macros/require.sem')
+
+extern comparisonProviderInterface: [compare:[int256, int256]:int256]
+
+def init():
+    self.owner = msg.sender
+
+def any():
+    require(msg.sender == self.owner or msg.sender == self)
+
+def initialize(comparor: address):
+    require(not self.initialized)
+    self.initialized = 1
+    self.comparor = comparor
+    return(1)
+
+def add(item: int256, hints: arr):
+    require(item != 0)
+
+    if (self.contains(item)):
+        self.remove(item)
+
+    # If this is the only node it is the head / tail and we can early out
+    if (self.count == 0):
+        self.head = item
+        self.tail = item
+        self.collection[item].exists = 1
+        self.count += 1
+        return(1)
+
+    # By default we begin the insertion search at the head
+    nodeValue = self.head
+
+    # If hints are provided (and valid) we will use the first still present value to begin the search
+    hintIndex = 0
+    while(hintIndex < len(hints) and nodeValue == self.head):
```

Comment:
My preferred way is to extract this into a helper method and use `return` as the break:
```
def getBestHint(hints: arr, item):
    hintIndex = 0
    while hintIndex < len(hints):
        hint = hints[hintIndex]
        if self.isValidHint(hint, item):
            return hint
        hintIndex += 1
    return self.head
```
As seen in the above, I would also extract out `isValidHint` into a helper method because I don't like compound if statements:
```
def isValidHint(hint, index):
    if not self.contains(hint):
        return 0
    if not self.comparor.compare(hint, item) == 1:
        return 0
    return 1
```

All of the above is style of course, so entirely up to you.